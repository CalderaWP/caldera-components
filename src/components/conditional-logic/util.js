/**
 * Get the current values for prepareConfigFields
 *
 * @param configFields
 */
export const reduceConfigFieldsToValues = (configFields) => {
	let values = {};
	configFields.forEach(field => {
		values[field.ID] = field.value
			? field.value
			: field.default
				? field.default
				: null;
	});
	return values;
};
