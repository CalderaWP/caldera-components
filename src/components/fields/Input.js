import React from 'react';
import PropTypes from 'prop-types';
import {
	valuePropType,
	onValueChangePropType
} from './propTypes';

export const Input = (props) => {
	function changeHandler(event){
		return props.onValueChange(event.target.value);
	}
	return (
		<input
			type={props.inputType ? props.inputType : 'text'}
			id={props.id}
			className={props.fieldClassName}
			aria-describedby={props.ariaDescribedbyAttr}
			required={props.isRequired}
			onChange={changeHandler}
			value={props.value}
		/>
	);

};

Input.propTypes = {
	id: PropTypes.string.isRequired,
	fieldClassName: PropTypes.string.isRequired,
	inputType: PropTypes.string,
	ariaDescribedbyAttr: PropTypes.string,
	isRequired: PropTypes.bool,
	value: valuePropType,
	onValueChange: onValueChangePropType
};

Input.defaultProps = {
	ariaDescribedbyAttr: '',
	isRequired: false
};