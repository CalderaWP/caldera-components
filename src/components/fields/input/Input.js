import React from 'react';
import {
	fieldPropTypes
} from '../propTypes';

export const Input = (props) => {
	function changeHandler(event){
		return props.onValueChange(event.target.value);
	}
	return (
		<input
			type={props.inputType}
			id={props.id}
			className={props.fieldClassName}
			aria-describedby={props.ariaDescribedbyAttr}
			required={props.isRequired}
			onChange={changeHandler}
			value={props.value}
		/>
	);

};

Input.propTypes = fieldPropTypes;

Input.defaultProps = {
	ariaDescribedbyAttr: '',
	isRequired: false,
	inputType: 'text'
};