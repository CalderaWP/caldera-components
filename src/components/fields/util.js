/**
 * Create the aria-describedby attribute or id attribute for the describing element
 *
 * @param {String} id Id attribute of input being described
 * @param {String} help Optional. Help text. If empty return is null.
 * @returns {String|null}
 */
export const  ariaDescribedbyAttr = (id, help = '') => {
	let ariaDescribedby = null;
	if (help.length) {
		ariaDescribedby = `${id}-description`;
	}
	return ariaDescribedby;
};


/**
 * Get HTML5 input types that are valid
 * @returns {string[]}
 */
export function getHtmlInputTypes() {
	return [
		'text',
		'email',
		'number',
		'date',
		'datetime',
		'password',
		'submit',
		'reset',
		'checkbox',
		'hidden'
	];
}

/**
 * Checks if a given input type is an acceptable HTML5 input type
 *
 * @param {String} type
 * @returns {boolean}
 */
export function isValidHtml5type(type) {
	return getHtmlInputTypes().includes(type);
}


/**
 * Remove a value from an array if present, if not present, add it
 *
 * @param {String|number} value Value to add or remove
 * @param {Array} array Array to mutate
 * @return {*}
 */
export const addOrRemoveFromArray = (value,array) => {
	const index = array.indexOf(value);
	if (index !== -1){
		array.splice(index, 1);
	}else{
		array.push( value );
	}
	return array;
};

/**
 * Cast a boolean or boolean like to a true or false
 *
 * @param Mixed} value Value to cast
 * @return {boolean}
 */
export const toBoolean =(value) => {
	switch(value){
	case true:
	case 'true':
	case 1:
	case '1':
	case 'on':
	case 'yes':
		return true;
	default:
		return false;
	}
};