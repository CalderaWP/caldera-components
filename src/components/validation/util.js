import getValidatorsFromConfigField from './getValidatorsFromConfigField';
import isValid from './isValid';
import isValidOrEmpty from './isValidOrEmpty';

/**
 * Find the value of a field by ID
 *
 * Returns null if field is not in collection
 *
 * @param {String} fieldId Field to look for
 * @param {Object} fieldValues Values to look for field in
 * @return {String|Array|boolean|number|null}
 */
export const findFieldValueInFieldValues = (fieldId, fieldValues) => {
	if (fieldValues.hasOwnProperty(fieldId)) {
		return fieldValues[fieldId];
	}

	return null;
};

/**
 * Helper function to curry validator functions
 *
 * Validators expect 1 value, not an object of values, this function solves that.
 *
 * @param {String} fieldId Field to look for
 * @param {Function} validator Validator function
 * @return {function(*=): *}
 */
function curryValidator(fieldId, validator) {
	return (fieldValues) => {
		return validator(findFieldValueInFieldValues(fieldId, fieldValues));
	};
}

/**
 * Create validation function based on type and required
 *
 * @param {Object} configField Field to create validations for
 * @param {Array} validators Current array of validators
 * @param {string} validationType Type of validators to add.
 * @return {Array}
 */
function addValidatorsForType(configField, validators, validationType) {
	if (configField.isRequired) {
		validators.push(
			curryValidator(
				configField.ID, isValid[validationType]
			)
		);
	} else {
		validators.push(
			curryValidator(
				configField.ID, isValidOrEmpty[validationType]
			)
		);
	}

	return validators;
}

/**
 * Adds sensible, automatic validation to fields -- if no validators have been added yet.
 *
 * @param {Object} configField
 * @return {*}
 */
export const addAutomaticValidators = (configField) => {
	let validators = getValidatorsFromConfigField(configField);
	if (0 !== validators.length) {
		return configField;
	}

	switch (configField.type) {
	case 'input':
		if( [
			'email',
			'url',
			'date',
			'number',
		].includes( configField.inputType ) ){
			validators = addValidatorsForType(configField, validators, configField.inputType);
		}else if( 'text' === configField.inputType){
			validators = addValidatorsForType(configField, validators, 'string');

		}

		else{
			validators = addValidatorsForType( configField, validators, 'anything' );
		}

		break;
	case 'select':
		validators = addValidatorsForType( configField, validators, 'stringOrNumber' );
		break;
	}


	return {
		...configField,
		validators
	};


};
