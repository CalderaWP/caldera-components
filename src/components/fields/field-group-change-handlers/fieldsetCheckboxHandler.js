import {addOrRemoveFromArray} from '../util';

/**
 * Change handler for a checkbox that is a part of a group
 * @param {Object} option Option that is being changed. Shape: `{value:String|number,label:String}`,
 * @param {Array|string|number} values Current values
 * @param {Function} onValueChange Function to call with updated value
 */
export const fieldsetCheckboxHandler = (option, values, onValueChange)  => {
	if (!Array.isArray(values)) {
		if ('string' === typeof values || 'number' === typeof  values) {
			values = [values];
		} else {
			values = [];

		}
	}
	values = addOrRemoveFromArray(option.value, values);
	onValueChange(values);
};