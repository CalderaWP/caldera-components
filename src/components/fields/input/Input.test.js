import renderer from 'react-test-renderer';
import React from 'react';
import {shallow} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {Input} from './Input';

Enzyme.configure({adapter: new Adapter()});
import {getHtmlInputTypes} from '../util';
import {fieldsetCheckboxHandler} from "../field-group-change-handlers/fieldsetCheckboxHandler";

describe('Input component', () => {
	describe('Input component props', () => {
		it('Passes its props', () => {
			const component = renderer.create(
				<Input
					id={'bags'}
					fieldClassName={'foo'}
					onValueChange={() => {
					}}
				/>
			);
			expect(component.toJSON()).toMatchSnapshot();
		});

		it('passes the value', () => {
			const wrapper = renderer.create(
				<Input
					id={'bags'}
					fieldClassName={'foo'}
					onValueChange={() => {
					}}
					value={'Sivans'}
				/>
			);
			expect(wrapper.toJSON()).toMatchSnapshot();
		});
	});

	describe('Input component rendering', () => {
		it('Has the correct id attribute', () => {
			const wrapper = shallow(
				<Input
					id={'bags'}
					fieldClassName={'foo'}
					onValueChange={() => {
					}}

				/>
			);
			expect(wrapper.find('#bags')).toHaveLength(1);
		});

		it('Has the correct class attribute', () => {
			const wrapper = shallow(
				<Input
					id={'bags'}
					fieldClassName={'foo'}
					onValueChange={() => {
					}}
				/>
			);
			expect(wrapper.find('.foo')).toHaveLength(1);
		});

		it('Has the value', () => {
			const wrapper = shallow(
				<Input
					id={'bags'}
					fieldClassName={'foo'}
					onValueChange={() => {
					}}
					value={'Sivans'}
				/>
			);
			expect(wrapper.find('input').prop('value')).toBe('Sivans');
		});

		it('Updates the value when changed', () => {
			let setValue = '';
			const wrapper = shallow(
				<Input
					id={'bags'}
					fieldClassName={'foo'}
					onValueChange={(newValue) => {
						setValue = newValue;
					}}
					value={'Sivans'}
				/>
			);
			wrapper.find('input').simulate('change', {target: {value: 'r3s'}});
			expect(setValue).toEqual('r3s');
		});

		describe('Disabled prop', () => {
			it('Does not disable by default', () => {
				const wrapper = shallow(
					<Input
						id={'i8'}
						fieldClassName={'foo'}
						onValueChange={() => {
						}}
						value={'Sivans'}
					/>
				);
				expect(wrapper.find('input').prop('disabled')).toBe(undefined);
			});

			it('Does  disable by default', () => {
				const wrapper = shallow(
					<Input
						id={'i8'}
						fieldClassName={'foo'}
						onValueChange={() => {
						}}
						value={'Sivans'}
						disabled={true}
					/>
				);
				expect(wrapper.find('input').prop('disabled')).toBe(true);
			});
		});
	});

	describe('Works for all HTML5 input types via inputType prop', () => {
		getHtmlInputTypes().forEach((type) => {
			it(`inputType prop of ${type} works`, () => {
				const wrapper = shallow(
					<Input
						id={'bags'}
						fieldClassName={'foo'}
						onValueChange={() => {
						}}
						value={'Sivans'}
						inputType={type}
					/>
				);
				expect(wrapper.find('input').prop('type')).toBe(type);
			});

		});
	});

	describe('Checkbox', () => {
		describe('props for checkboxes', () => {
			it('Can be a checkbox', () => {
				const component = renderer.create(
					<Input
						id={'check1'}
						fieldClassName={'check1'}
						onValueChange={() => {
						}}
						value={'true'}
						inputType={'checkbox'}
					/>
				);
				expect(component.toJSON()).toMatchSnapshot();
			});

			it('Has checkbox type', () => {
				const wrapper = shallow(
					<Input
						id={'check2'}
						fieldClassName={'check1'}
						onValueChange={() => {
						}}
						value={'true'}
						inputType={'checkbox'}
					/>
				);
				expect(wrapper.find('input').prop('type')).toEqual('checkbox');
			});

			it('Sets defaultChecked prop to true', () => {
				let value = true;
				const wrapper = shallow(
					<Input
						id={'check5'}
						fieldClassName={'check1'}
						onValueChange={(newValue) => {
							value = newValue;
						}}
						value={value}
						inputType={'checkbox'}
					/>
				);
				expect(wrapper.find('input').prop('defaultChecked')).toEqual(true);
			});

			it('Sets defaultChecked prop to false', () => {
				let value = false;
				const wrapper = shallow(
					<Input
						id={'check6'}
						fieldClassName={'check1'}
						onValueChange={(newValue) => {
							value = newValue;
						}}
						value={value}
						inputType={'checkbox'}
					/>
				);
				expect(wrapper.find('input').prop('defaultChecked')).toEqual(false);
			});

		});

	});

});




