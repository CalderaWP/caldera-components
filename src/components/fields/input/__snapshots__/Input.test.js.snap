// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Input component props Passes its props 1`] = `
<input
  aria-describedby=""
  className="foo"
  id="bags"
  onChange={[Function]}
  required={false}
  type="text"
/>
`;

exports[`Input component Input component props passes the value 1`] = `
<input
  aria-describedby=""
  className="foo"
  id="bags"
  onChange={[Function]}
  required={false}
  type="text"
  value="Sivans"
/>
`;
